editor/plugins/control_editor_plugin.cpp

```diff
@@ -594,400 +822,208 @@ Vector2 ControlEditorToolbar::_position_to_anchor(const Control *p_control, Vect
+
+			Container *parent_container = Object::cast_to<Container>(control->get_parent());
+
+			Vector<int> container_h_flags = parent_container->get_allowed_size_flags_horizontal();
+			Vector<SizeFlags> tmp_flags = allowed_h_flags.duplicate();
+			for (int i = 0; i < allowed_h_flags.size(); i++) {
+				if (!container_h_flags.has((int)allowed_h_flags[i])) {
+					tmp_flags.erase(allowed_h_flags[i]);
+				}
+			}
+			allowed_h_flags = tmp_flags;
+
+			Vector<int> container_v_flags = parent_container->get_allowed_size_flags_vertical();
+			tmp_flags = allowed_v_flags.duplicate();
+			for (int i = 0; i < allowed_v_flags.size(); i++) {
+				if (!container_v_flags.has((int)allowed_v_flags[i])) {
+					tmp_flags.erase(allowed_v_flags[i]);
+				}
+			}
+			allowed_v_flags = tmp_flags;
 		}
+	}
+
+	// Set general toolbar visibility.
+	set_visible(has_controls);
+
+	// Set anchor tools visibility.
+	if (has_controls && (!has_control_parents || !has_container_parents)) {
+		anchors_button->set_visible(true);
+		anchor_mode_button->set_visible(true);
 
-		nb_valid_controls++;
-		if (control->get_meta("_edit_use_anchors_", false)) {
-			nb_anchors_mode++;
+		// Update anchor mode.
+		int nb_valid_controls = 0;
+		int nb_anchors_mode = 0;
+
+		List<Node *> selection = editor_selection->get_selected_node_list();
+		for (Node *E : selection) {
+			Control *control = Object::cast_to<Control>(E);
+			if (!control) {
+				continue;
+			}
+			if (Object::cast_to<Container>(control->get_parent())) {
+				continue;
+			}
+
+			nb_valid_controls++;
+			if (control->get_meta("_edit_use_anchors_", false)) {
+				nb_anchors_mode++;
+			}
 		}
+
+		anchors_mode = (nb_valid_controls == nb_anchors_mode);
+		anchor_mode_button->set_pressed(anchors_mode);
+	} else {
+		anchors_button->set_visible(false);
+		anchor_mode_button->set_visible(false);
+		anchor_mode_button->set_pressed(false);
 	}
 
-	anchors_mode = (nb_valid_controls == nb_anchors_mode);
-	anchor_mode_button->set_pressed(anchors_mode);
+	// Set container tools visibility.
+	if (has_controls && (!has_control_parents || has_container_parents)) {
+		containers_button->set_visible(true);
 
-	if (nb_controls > 0) {
-		set_physics_process(true);
+		// Update allowed size flags.
+		if (has_container_parents) {
+			container_h_picker->set_allowed_flags(allowed_h_flags);
+			container_v_picker->set_allowed_flags(allowed_v_flags);
+		} else {
+			Vector<SizeFlags> allowed_all_flags = {
+				SIZE_SHRINK_BEGIN,
+				SIZE_SHRINK_CENTER,
+				SIZE_SHRINK_END,
+				SIZE_FILL,
+				SIZE_EXPAND,
+			};
+
+			container_h_picker->set_allowed_flags(allowed_all_flags);
+			container_v_picker->set_allowed_flags(allowed_all_flags);
+		}
 	} else {
-		set_physics_process(false);
-		set_visible(false);
+		containers_button->set_visible(false);
 	}
 }
 
 void ControlEditorToolbar::_notification(int p_what) {
 	switch (p_what) {
 		case NOTIFICATION_ENTER_TREE:
 		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
```

Comment:
I just noticed now, but the EDITOR_SETTINGS_CHANGED is for handling theme update, no? Couldn't it be THEME_CHANGED?